# Lista de Exercícios I - Estatística I

<br>

__Obs.:__ A presença da aula de exercícios está atrelada a entrega deles. Ou seja, caso o aula não queria ficar durante a aula, não tem problema desde que entregue os exercícios dentro do prazo estipulado.

__Prazo:__ 05/03

__1)__ A probabilidade de dado honesto é conhecido e igual a $1/6$, mas quando vamos testar isso com um dado físico, não necessariamente vai cair uma vez cada face do dado (se eu lançar ele 6 vezes). A ideia é que conforme fazermos vários testes a tendência é que a probabilidade de cada face irá convergir pro valor esperado. 

Vamos testar este conceito utilizando da programação seguintes os itens abaixo:]

__A)__ Desenvolva uma função que receba um valor n, onde n é a quantidade de elementos e o objetivo da função é retorna uma _Series_ com n elementos sendo esses elementos números variando de 1 e 6;

__B)__ Teste a função para n igual a 10, 100, 1.000, 10.000 e 100.000 casos e para cada um deles calcule a proporção de vezes que aparece cada número;

__DIca.:__ utilize a função do _Pandas_ `.value_counts`

__C)__ O que podermos dizer sobre a afirmação do enunciado, a tendência é que as probabilidades irâo convergir para $1/6$?

In [ ]:
""" A) Desenvolva uma função que receba um valor n, onde n é a quantidade de elementos e 
       o objetivo da função é retorna uma Series com n elementos sendo esses elementos números variando de 1 e 6;"""

import pandas as pd
import numpy as np

def dado(n):

    num_dado = np.random.randint(1, 7, n)
    return pd.Series(num_dado)

dado(5)

0    3
1    2
2    6
3    5
4    6
dtype: int64

In [ ]:
""" B) Teste a função para n igual a 10, 100, 1.000, 10.000 e 100.000 casos e 
       para cada um deles calcule a proporção de vezes que aparece cada número """

n_valores = [10, 100, 1000, 10000, 100000]

for n in n_valores:
    rolagem = dado(n)
    count = rolagem.value_counts(normalize=True)
    print(f"n = {n}\n{count}\n")


n = 10
6    0.3
2    0.2
4    0.2
1    0.2
5    0.1
dtype: float64

n = 100
2    0.21
6    0.20
1    0.17
3    0.15
5    0.14
4    0.13
dtype: float64

n = 1000
4    0.186
1    0.174
5    0.166
6    0.161
3    0.157
2    0.156
dtype: float64

n = 10000
5    0.1739
6    0.1665
2    0.1662
3    0.1647
1    0.1644
4    0.1643
dtype: float64

n = 100000
4    0.16849
5    0.16693
3    0.16645
1    0.16629
6    0.16606
2    0.16578
dtype: float64



C) O que podermos dizer sobre a afirmação do enunciado, a tendência é que as probabilidades irâo convergir para  1/6 ?


De fato, a medida que vamos aumentando a quantidade de lançamentos vemos
uma maior uniformidade na proporção dos resultados, tendendo para ~16,7% 
(0.1666...)
___


__2)__ Um programa de computador gera uma pontuação aleatória para cada usuário, sendo que:

- Em 1/2 dos casos, gera 0.5;
- Em 1/4 dos casos, gera 0.25;
- Em 1/8 dos casos, gera 0.125.

E essa lógica segue assim por diante. Resolva os itens a seguir:

__A)__ Calcule o __valor esperado__ utilizando apenas os 3 primeiros casos;

__B)__ Faça um _loop_ para calcular o __valor esperado__ para 100 casos e avalie se só com os 3 primeiros casos temos uma boa aproximação ou não.

In [ ]:
# A) Calcule o valor esperado utilizando apenas os 3 primeiros casos

valor_esperado = (1/2) * 0.5 + (1/4) * 0.25 + (1/8) * 0.125
valor_esperado

0.328125

In [ ]:
""" B) Faça um loop para calcular o valor esperado para 100 casos 
   e avalie se só com os 3 primeiros casos temos uma boa aproximação ou não. """

valor = 0 # Só para zerar a variavel caso rode mais de uma vez

for i in range(1,101):
  resultado = 1/(2**i)
  valor += resultado*resultado
print(f"Valor acumulado = {valor}\nProporção dos 3 primeiros e os 100 casos: {valor_esperado/valor}")

Valor acumulado = 0.3333333333333333
Proporção dos 3 primeiros e os 100 casos: 0.984375


O valor aculumado para os 100 casos é de 0.333... e como o valor dos 3 primeiros
casos é de 0.328125, ao fazermos a proporção entre eles temos o resultado de 0.984375 (98% de precisão).

Nesse caso, o valor esperado obtido com 100 casos é muito próximo do valor esperado 
calculado apenas com os três primeiros casos, sugerindo que a aproximação é excelente!
___

__3)__ Suponha que o limite de uma estrada é de 80km/h e, a partir da medição por radar, sabe-se que nessa rodovia a velocidade dos veículo $X$ segue uma distribuição normal com média de 72km/h e desvio padrão de 8 km/h.
 - Qual a probabilidade de um veículo ser multado?
 - Além da multa, se um veículo ultrapassar o limite em 30km/h, o motorista perde a carteira. Considerando apenas os infratores de velocidade, qual a probabilidade dele perder a carteira?
    

In [ ]:
# Qual a probabilidade de um veículo ser multado?

from scipy.stats import norm

media = 72
desvio_padrao = 8
probabilidade_multa = 1 - norm.cdf(81, media, desvio_padrao)
print(f"A probabilidade de um veículo ser multado é de: {probabilidade_multa:.2%}")

A probabilidade de um veículo ser multado é de: 13.03%


In [ ]:
perder_carteira = 1 - norm.cdf(110, media, desvio_padrao)
print(f"A probabilidade do motorista perder a carteira é de: {perder_carteira:.6%}")

A probabilidade do motorista perder a carteira é de: 0.000102%


O resultado é uma probabilidade muito próxima de zero, indicando que é muito improvável que um veículo infrator atinja essa velocidade e perca a carteira.
___

__4)__ Suponha que a probabilidade de uma criança nascer loira é de 1/4. Se houverem 6 crianças na família, qual é a probabilidade de 3 dele serem loiros?

In [ ]:
#P(X=k) = C(n,k) * p^k * (1-p)^(n-k)

def fac(n):
    if n == 0:
        return 1
    else:
        return n * fac(n-1)

# P(X=3) = C(6,3) * (1/4)^3 * (3/4)^6-3

c = fac(6) / (fac(3) * fac(6-3))
probabilidade = c * (1/4)**3 * (3/4)**3

print(f"Portanto, a probabilidade de termos exatamente 3 crianças loiras em uma de 6 crianças é de aproximadamente:\n {probabilidade:.2%}")

Portanto, a probabilidade de termos exatamente 3 crianças loiras em uma de 6 crianças é de aproximadamente:
 13.18%


__5)__ Se a probabilidade de atingir um alvo com um único disparo é de 0.3. Qual é a probabilidade de, em 4 disparos, atingirmos o alvo ao menos 3 vezes?


In [ ]:
# Probabilidade de atingir 3x em 4 tiros
# P(X=3) = C(4,3) * (0.3)^3  * (1-0,3)^4-3

c3 = fac(4) / (fac(3) * fac(4-3))
probabilidade3 = c3 * (0.3)**3 * (1-0.3)**1

# Probabilidade de atingir 4x em 4 tiros
# P(X=4) = comb(4,4) * (0.3)^4 * (0.7)^4-4
c4 = fac(4) / (fac(4) * fac(4-4))
probabilidade4 = c4 * (0.3)**4 * (0.7)**0

print(f"A probabilidade de acerto é de aproximadamente: {probabilidade3 + probabilidade4:.2%}")

A probabilidade de acerto é de aproximadamente: 8.37%


__6)__ Suponha que o número de meteoros registrados por um sistema de radares num intervalo qualquer de 30 segundos é de 1.81 em média. Assumindo que os meteoros aparecem aleatoriamente e de forma independente. Usando a distribuição de Poisson determine:
 - A probabilidade de que nenhum meteoro seja encontrado num intervalo de 1 minuto.
 - A probabilidade encontrarmos de 5 a 8 meteoros em 2 minutos.

In [ ]:
# A probabilidade de que nenhum meteoro seja encontrado num intervalo de 1 minuto

def poisson(lbd, k):
    return (np.exp(-lbd)*(lbd**k))/fac(k)
# Como a probabilidade é de 1.81 acada 30s, vamos duplicar por se tratar de 1 min
lbd =  1.81 * 2

# Como queremos 0 meteoros
k = 0

print(f"Probabilidade de não encontrar nenhum meteoro em um minuto: {poisson(lbd,k):.4f} ou {poisson(lbd,k):.2%}")

Probabilidade de não encontrar nenhum meteoro em um minuto: 0.0268 ou 2.68%


In [ ]:
# Forma importando a biblioteca

# from scipy.stats import poisson

# media = 1.81 * 2
# prob_zero_meteoro = poisson.pmf(0, media)

# print("Probabilidade de não encontrar nenhum meteoro em um minuto:", prob_zero_meteoro)

In [ ]:
# Probabilidade encontrarmos de 5 a 8 meteoros em 2 minutos.
lbd = 1.81 * 4

resultado = 0 # Só para zerar a variavel caso rode mais de uma vez

#Criando um loop para soma das probabilidades de 5 a 8 meteoros
for k in range(5, 9):
  resultado += poisson(lbd,k)

print(f"Probabilidade de encontrar de 5 a 8 meteoros em 2 minutos: {resultado:.4f} ou {resultado:.2%}")


Probabilidade de encontrar de 5 a 8 meteoros em 2 minutos: 0.5451 ou 54.51%


In [ ]:
# Forma importando a biblioteca

# from scipy.stats import poisson
# media = 1.81 * 4
# prob_5_a_8_meteoros = sum(poisson.pmf(i, media) for i in range(5, 9))

# print("Probabilidade de encontrar de 5 a 8 meteoros em 2 minutos:", prob_5_a_8_meteoros)

__7)__ O número médio de jornais vendidos numa banca é 5 jornais por minuto. Qual é a probabilidade de vender ao menos 1 jornal em um minuto?

In [ ]:
lbd = 5

# Como a probabilidade é de vender pelo menos 1 jornal, então fazemos a proporção
# inversa, que no caso seria não vender nenhum e retirar de 100%.
k = 0
print(f"A probabilidade de vender ao menos 1 jornal em um minuto é de: {1 - poisson(lbd,k):.4f} ou {1 - poisson(lbd,k):.2%}")

A probabilidade de vender ao menos 1 jornal em um minuto é de: 0.9933 ou 99.33%


__8)__ (Monty-Hall) Em um programa de auditório, havia 3 portas (A, B, C), dentro da qual uma continha um prêmio. O participante escolheu a porta A e antes de abrir a porta escolhida o apresentador abriu a porta C, na qual não havia nada. O participante aumentará a probabilidade de acertar a porta caso mude a escolha dele da porta A para a B? Demonstre a sua resposta utilizando o Teorema de Bayes? 

Podemos calcular as probabilidades condicionais usando o Teorema de Bayes:

P(A|B) = P(B|A) * P(A) / P(B)
P(A|C) = P(C|A) * P(A) / P(C)

Sabemos que P(B|A) = 1/2, porque se o prêmio está atrás da porta A, o apresentador pode abrir a porta B ou a porta C com igual probabilidade. Além disso, P(C|A) = 1, porque se o prêmio está atrás da porta A, o apresentador deve abrir a porta que não tem o prêmio.

Assumindo que o participante escolheu a porta A inicialmente, temos P(A) = 1/3 e P(B) = 1/2, porque o apresentador só pode abrir a porta B se o prêmio não estiver atrás da porta A.

Substituindo esses valores, obtemos:

P(A|B) = (1/2) * (1/3) / (1/2) = 1/3

P(A|C) = 1 * (1/3) / (1/2) = 2/3

Portanto, se o participante mudar de escolha para a porta B, ele tem uma probabilidade maior de acertar o prêmio, que é de 2/3, em vez de 1/3 se ele mantiver a escolha inicial na porta A.

In [ ]:
# Probabilidade inicial de cada porta conter o prêmio
p_a = p_b = p_c = 1/3

# Após a porta C ser aberta e não haver o prêmio
p_c_sem_premio = 1
p_a_sem_premio = p_a * p_c_sem_premio / (p_a * p_c_sem_premio + p_b * p_c_sem_premio + p_c * p_c_sem_premio)
p_b_sem_premio = p_b * p_c_sem_premio / (p_a * p_c_sem_premio + p_b * p_c_sem_premio + p_c * p_c_sem_premio)

# Probabilidade de o prêmio estar em cada porta após a porta C ser aberta e não haver o prêmio
p_a_com_troca = p_a_sem_premio 
p_b_com_troca = (1 - p_a_sem_premio) 

print("Probabilidade do prêmio estar na porta A sem a troca:", p_a_com_troca)
print("Probabilidade do prêmio estar na porta B com a troca:", p_b_com_troca)

Probabilidade do prêmio estar na porta A sem a troca: 0.3333333333333333
Probabilidade do prêmio estar na porta B com a troca: 0.6666666666666667
